kbncontent.go

```diff
@@ -0,0 +1,248 @@
+}
+
+func getVisType(doc interface{}, soType string) (string, error) {
+	if soType != "visualization" {
+		return "", nil
+	}
+
+	if attrType, err := jsonpath.Get("$.attributes.type", doc); err == nil {
+		return attrType.(string), nil
+	}
+
+	if visStateType, err := jsonpath.Get("$.attributes.visState.type", doc); err == nil {
+		return visStateType.(string), nil
+	}
+
+	// by-value dashboard case
+	if embeddableType, err := jsonpath.Get("$.embeddableConfig.savedVis.type", doc); err == nil {
+		return embeddableType.(string), nil
+	}
+
+	return "", nil
+}
+
+func isTSVB(visType string) bool {
+	return visType == "metrics"
+}
+
+func getTSVBType(doc interface{}, visType string) (string, error) {
+	if !isTSVB(visType) {
+		return "", nil
+	}
+
+	// saved object case
+	if result, err := jsonpath.Get("$.attributes.visState.params.type", doc); err == nil {
+		return result.(string), nil
+	}
+
+	// by-value dashboard panel case
+	if result, err := jsonpath.Get("$.embeddableConfig.savedVis.params.type", doc); err == nil {
+		return result.(string), nil
+	}
+
+	return "", nil
+}
+
+func getVisTitle(doc interface{}, soType string) (string, error) {
+	if soType != "visualization" {
+		return "", nil
+	}
+
+	if title, err := jsonpath.Get("$.attributes.title", doc); err == nil {
+		return title.(string), nil
+	}
+
+	if title, err := jsonpath.Get("$.title", doc); err == nil {
+		return title.(string), nil
+	}
+
+	// by-value dashboard case
+	if title, err := jsonpath.Get("$.embeddableConfig.savedVis.title", doc); err == nil {
+		return title.(string), nil
+	}
+
+	return "", nil
+}
+
+// Attaches domain knowledge as well as information from within the document
+func attachMetaInfo(desc *VisDesc) {
+	if result, err := getVisType(desc.Doc, desc.SoType); err == nil {
+		desc.Type = result
+	}
+
+	if result, err := getVisTitle(desc.Doc, desc.SoType); err == nil {
+		desc.Title = result
+	}
+
+	if result, err := getTSVBType(desc.Doc, desc.Type); err == nil {
+		desc.TSVBType = result
+	}
+
+	if isTSVB(desc.Type) {
+		desc.SemanticType = desc.TSVBType
+	} else {
+		desc.SemanticType = desc.Type
+	}
+
+	desc.IsLegacy = isLegacy(desc.SoType, desc.Type)
+}
+
+func deserializeSubPaths(doc objx.Map) (map[string]interface{}, error) {
+	uiState := doc.Get("attributes.uiStateJSON")
+	if uiState.IsStr() {
+		if parsed, err := objx.FromJSON(uiState.Str()); err == nil {
+			doc.Set("attributes.uiStateJSON", parsed)
+		}
+	}
+
+	visState := doc.Get("attributes.visState")
+	if visState.IsStr() {
+		if parsed, err := objx.FromJSON(visState.Str()); err == nil {
+			doc.Set("attributes.visState", parsed)
+		}
+	}
+
+	searchSource := doc.Get("attributes.kibanaSavedObjectMeta.searchSourceJSON")
+	if searchSource.IsStr() {
+		if parsed, err := objx.FromJSON(searchSource.Str()); err == nil {
+			doc.Set("attributes.kibanaSavedObjectMeta.searchSourceJSON", parsed)
+		}
+	}
```

Comment:
We can probably use a loop here over the related keys, and we probably want to return the errors happening here.
```suggestion
	jsonFields := []string{
		"attributes.uiStateJSON",
		"attributes.visState",
		"attributes.kibanaSavedObjectMeta.searchSourceJSON",
	}
	for _, fieldName := range jsonFields {
		field := doc.Get(fieldName)
		if !field.IsStr() {
			continue
		}
		parsed, err := objx.FromJSON(field.Str())
		if err != nil {
			return nil, fmt.Errorf("failed to decode embedded json in %q: %w", fieldName, err)
		}
		doc.Set(fieldName, parsed)
	}
```

Reply:
Beautiful.